libraries/AP_NavEKF/EKFGSF_yaw.cpp

```diff
@@ -0,0 +1,678 @@
+	AHRS[mdl_idx].R[0][0] = R_prev[0][0] * cos_yaw - R_prev[1][0] * sin_yaw;
+	AHRS[mdl_idx].R[0][1] = R_prev[0][1] * cos_yaw - R_prev[1][1] * sin_yaw;
+	AHRS[mdl_idx].R[0][2] = R_prev[0][2] * cos_yaw - R_prev[1][2] * sin_yaw;
+	AHRS[mdl_idx].R[1][0] = R_prev[0][0] * sin_yaw + R_prev[1][0] * cos_yaw;
+	AHRS[mdl_idx].R[1][1] = R_prev[0][1] * sin_yaw + R_prev[1][1] * cos_yaw;
+	AHRS[mdl_idx].R[1][2] = R_prev[0][2] * sin_yaw + R_prev[1][2] * cos_yaw;
+
+}
+
+void EKFGSF_yaw::initialise()
+{
+	memset(&GSF, 0, sizeof(GSF));
+	vel_fuse_running = false;
+    vel_data_updated = false;
+    run_ekf_gsf = false;
+
+	memset(&EKF, 0, sizeof(EKF));
+	const float yaw_increment = M_2PI / (float)N_MODELS_EKFGSF;
+	for (uint8_t mdl_idx = 0; mdl_idx < N_MODELS_EKFGSF; mdl_idx++) {
+		// evenly space initial yaw estimates in the region between +-Pi
+		EKF[mdl_idx].X[2] = -M_PI + (0.5f * yaw_increment) + ((float)mdl_idx * yaw_increment);
+
+		// All filter models start with the same weight
+		GSF.weights[mdl_idx] = 1.0f / (float)N_MODELS_EKFGSF;
+
+		// Take state and variance estimates direct from velocity sensor
+		EKF[mdl_idx].X[0] = vel_NE[0];
+		EKF[mdl_idx].X[1] = vel_NE[1];
+		EKF[mdl_idx].P[0][0] = sq(fmaxf(vel_accuracy, 0.5f));
+		EKF[mdl_idx].P[1][1] = EKF[mdl_idx].P[0][0];
+
+		// Use half yaw interval for yaw uncertainty as that is the maximum that the best model can be away from truth
+        GSF.yaw_variance = sq(0.5f * yaw_increment);
+		EKF[mdl_idx].P[2][2] = GSF.yaw_variance;
+	}
+}
+
+// returns the probability of a selected model output assuming a gaussian error distribution
+float EKFGSF_yaw::gaussianDensity(const uint8_t mdl_idx) const
+{
+	const float t2 = EKF[mdl_idx].S[0][0] * EKF[mdl_idx].S[1][1];
+	const float t5 = EKF[mdl_idx].S[0][1] * EKF[mdl_idx].S[1][0];
+	const float t3 = t2 - t5; // determinant
+	float t4; // determinant inverse
+	if (fabsf(t3) > 1e-6f) {
+		t4 = 1.0f/t3;
+	} else {
+		t4 = 1.0f/t2;
+	}
+
+	// inv(S)
+	float invMat[2][2];
+	invMat[0][0] =   t4 * EKF[mdl_idx].S[1][1];
+	invMat[1][1] =   t4 * EKF[mdl_idx].S[0][0];
+	invMat[0][1] = - t4 * EKF[mdl_idx].S[0][1];
+	invMat[1][0] = - t4 * EKF[mdl_idx].S[1][0];
+
+ 	// inv(S) * innovation
+	float tempVec[2];
+	tempVec[0] = invMat[0][0] * EKF[mdl_idx].innov[0] + invMat[0][1] * EKF[mdl_idx].innov[1];
+	tempVec[1] = invMat[1][0] * EKF[mdl_idx].innov[0] + invMat[1][1] * EKF[mdl_idx].innov[1];
+
+	// transpose(innovation) * inv(S) * innovation
+	float normDist = tempVec[0] * EKF[mdl_idx].innov[0] + tempVec[1] * EKF[mdl_idx].innov[1];
+
+    // convert from a normalised variance to a probability assuming a Gaussian distribution
+	normDist = expf(-0.5f * normDist);
+	normDist *= sqrtf(t4)/ M_2PI;
+	return normDist;
+}
+
+bool EKFGSF_yaw::getLogData(float *yaw_composite, float *yaw_composite_variance, float yaw[N_MODELS_EKFGSF], float innov_VN[N_MODELS_EKFGSF], float innov_VE[N_MODELS_EKFGSF], float weight[N_MODELS_EKFGSF])
```

Comment:
I think EKFGSF_Yaw object should do it's own logging. What rate does it need to log at? Maybe use AP::logger().Write() at 10Hz or so?